README.md
systemd-cron-c
==============

## What's it?

This is a compatibility layer for crontab-to-systemd timers framework. It works by parsing
crontab and anacrontab files from usual places like `/etc/crontab` and `/var/spool/cron`
and generating systemd timers and services. You can use `cron.target` as a single control
point for the generated units.

It's intended to be drop-in replacement for all cron implementations.

## More

I'm mostly interrested in learning C at a slow pace;
so users are better off using the supported [Python][]
version or the [Rust][] rewrite for now.

[Python]: https://github.com/systemd-cron/systemd-cron
[Rust]: https://github.com/systemd-cron/systemd-cron-next


debian/changelog
systemd-cron (2.0~20230730) UNRELEASED; urgency=low

  * C rewrite

 -- Alexandre Detiste <alexandre.detiste@gmail.com>  Sat, 29 Jul 2023 16:31:04 +0200


mail_on_failure.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>
#include <sys/wait.h>
#include <string.h>

char* read_systemd_variable(const char *unit, const char *property) {
	int filedes[2];
	if (pipe(filedes) == -1) {
		perror("pipe");
		exit(1);
	}

	pid_t pid = fork();
	if (pid == -1) {
		perror("fork");
		exit(1);
	} else if (pid == 0) {
		while ((dup2(filedes[1], STDOUT_FILENO) == -1) && (errno == EINTR)) {}
		close(filedes[1]);
		close(filedes[0]);
		char *arg;
		asprintf(&arg, "--property=%s", property);
		execl("/usr/bin/systemctl", "systemctl", "show", unit, arg, NULL);
		perror("execl");
		_exit(1);
	}
	close(filedes[1]);
	char buffer[4096];
	memset(buffer, '\0', sizeof(buffer));
	read(filedes[0], buffer, sizeof(buffer));
	close(filedes[0]);
	wait(0);

	char *result;
	asprintf(&result, "%s", buffer + strlen(property) + 1);
	result[strlen(result)-1] = '\0';
	return result;
}

int main(int argc, char *argv[]) {
	if (argc != 2) {
		fprintf(stderr, "Usage: %s <unit>\n", argv[0]);
		exit(1);
	}
        struct stat sb;
        if (stat("/usr/sbin/sendmail", &sb) == -1) {
		fprintf(stderr, "<3>can't send error mail for %s without a MTA\n", argv[1]);
		exit(1);
	}

	char *mailto = NULL;

	char *environment;
	environment = read_systemd_variable(argv[1], "Environment");
	printf("[%s]\n", environment);
	if (strlen(environment)) {
		// leak
		mailto = strstr(environment, "MAILTO=");
		if (mailto) {
			mailto += strlen("MAILTO=");

			char *p;
                        p = strchr(mailto, ' ');
			if (p)
			    p[0] = '\0';

			if (!strlen(mailto)) {
				free(environment);
				exit(0);
			}
		}
	}

	if (!mailto) {
		mailto = read_systemd_variable(argv[1], "User");
		if(!strlen(mailto)) {
			free(mailto);
			mailto = strdup("root");
		}
	}
	printf("Mailto: [%s]\n", mailto);

/*
hostname = os.uname()[1]

body = "From: root (systemd-cron)\n"
body += "To: " + mailto + "\n"
body += "Subject: [" + hostname + "] job " + job  + " failed\n"
body += "MIME-Version: 1.0\n"
body += "Content-Type: text/plain; charset=UTF-8\n"
body += "Content-Transfer-Encoding: 8bit\n"
body += "Auto-Submitted: auto-generated\n"
body += "\n"

try:
    systemctl = subprocess.check_output(['systemctl','status',job], universal_newlines=True)
except subprocess.CalledProcessError as e:
    if e.returncode != 3:
        raise
    else:
        body += e.output

p = Popen(['sendmail','-i','-B8BITMIME',mailto], stdin=PIPE)
p.communicate(bytes(body, 'UTF-8'))
*/

	free(environment);
}


crontab-generator.c
/*-*- Mode: C; c-basic-offset: 8; indent-tabs-mode: nil -*-*/

/***
  This file is part of systemd.

  Copyright 2013 Shawn Landden

  systemd is free software; you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation; either version 2.1 of the License, or
  (at your option) any later version.

  systemd is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with systemd; If not, see <http://www.gnu.org/licenses/>.
***/

#include <errno.h>
#include <stdio.h>
#include <unistd.h>
#include <dirent.h>
#include <stdbool.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#include "log.h"
#include "util.h"
#include "fileio.h"

#ifndef CRONTAB_PATH
#define CRONTAB_PATH "/etc/crontab"
#endif

#ifndef CRONTAB_DIR_PATH
#define CRONTAB_DIR_PATH "/etc/cron.d/"
#endif

#ifndef SYSTEMD_GENERATOR_PATH
#define SYSTEMD_GENERATOR_PATH "/run/systemd/generator"
#endif

const char *daysofweek[] = {"Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"};
const char *iscrontabline = "0123456789*";
const char *isdow = "0123456";

static const char *arg_dest = "/tmp";

static int parse_dow(char *dow, char *dows){
        char c[2] = {"\0\0"};

        assert(dow);

        dows[0] = '\0';

        for(unsigned s = 0;s < strlen(dow);s++) {
                c[0] = dow[s];
                if (strchr(isdow, c[0]))
                        strncpy(dows + strlen(dows), daysofweek[atoi(&c[0])], 128 - strlen(dows));
                else {
                        int l = strlen(dows);
                        if (l < 128) {
                                dows[l] = c[0];
                                dows[l+1] = '\0';
                        }
                }
        }

        return 0;
}

/*                             system means has a user field */
static int parse_crontab_entry(bool system, char *line, const char *context, char *usertab) {
        int r = 0;
        int commandloc = 0;
        char m[8], h[8], dom[8], mon[8], dow[8], user[64], command[1024];
        _cleanup_free_ char *timer = NULL, *timerfn = NULL, *service = NULL, *servicefn = NULL;
        char dows[128];
        char *p;

        assert(line);
        assert(context);

        if (system)
                r = sscanf(line, "%s %s %s %s %s %s %n", m, h, dom, mon, dow, user, &commandloc);
        else
                r = sscanf(line, "%s %s %s %s %s %n", m, h, dom, mon, dow, &commandloc);
        if (r == EOF)
                return -errno;

        p = strchr(&line[commandloc], '\n');
        p[0] = '\0';
        strcpy(&command[0], &line[commandloc]);
        if (user == NULL)
                strcpy(usertab, user);

        r = parse_dow(dow, &dows[0]);
        if (r < 0)
                return r;

        asprintf(&timerfn, SYSTEMD_GENERATOR_PATH "/cron-%s.timer", context);
        asprintf(&timer,"# Automatically generated by systemd-cron-generator\n"
                "# Was: %s\n"
                "[Unit]\n"
                "Description=[Cron] %s\n"
                "PartOf=cron.target\n"
                "RefuseManualStart=true\n"
                "RefuseManualStop=true\n"
                "\n"
                "[Timer]\n"
                "OnCalendar=%s *-%s-%s %s:%s\n"
                "Unit=cron-%s.service", line, context, dows, mon, dom, h, m, context);

        asprintf(&servicefn, SYSTEMD_GENERATOR_PATH "/cron-%s.service", context);
        asprintf(&service,"# Automatically generated by systemd-cron-generator\n"
                "# Was: %s\n"
                "[Unit]\n"
                "Description=[Cron] %s\n"
                "RefuseManualStart=true\n"
                "RefuseManualStop=true\n"
                "\n"
                "[Service]\n"
                "Type=oneshot\n"
                "User=%s"
                "ExecStart=/bin/sh -c \"%s\"\n", line, context, user, command);
        if (!timer || !timerfn || !service || !servicefn)
                return -log_oom();

        r = write_string_file(timerfn, timer);
        if (r < 0)
                return r;

        r = write_string_file(servicefn, service);
        if (r < 0)
                return r;

        return r;
}

static int parse_crontab(const char *filename) {
        int r = 0;
        _cleanup_fclose_ FILE *fp = NULL;
        char l[1024];
        _cleanup_free_ char *context = NULL;

        assert(filename);

        fp = fopen(filename, "r");
        if (!fp) {
                log_error("fopen() failed: %m");
                return -errno;
        }

        for (int ln = 0;l != NULL; ln++) {
                char *f;
                f = fgets(l, sizeof(l), fp);
                if (!f)
                        break;
                if (strchr(iscrontabline,l[0])) {
                        asprintf(&context, "%s-l%d", strrchr(filename, '/') + 1, ln);
                        r = parse_crontab_entry(true, (char *)&l, context, NULL);
                        if (r < 0)
                                return r;
                }
        }

        return r;
}


int main(int argc, char *argv[]) {
        int r = 0;
	DIR *dirp;
	struct dirent *dent;

        if (argc > 1 && argc != 4) {
                log_error("This program takes three or no arguments.");
                return EXIT_FAILURE;
        }

        if (argc > 1)
                arg_dest = argv[1];

        log_set_target(LOG_TARGET_SAFE);
        log_parse_environment();
        log_open();

        umask(0022);

        r = parse_crontab(CRONTAB_PATH);
	if (r < 0) {
                log_error("%m");
		goto exit;
        }

	dirp = opendir(CRONTAB_DIR_PATH);
	if (dirp == NULL) {
		log_error("opendir() failed: %m");
		r = -errno;
		goto exit;
	}

	while (1) {
                dent = readdir(dirp);
                if (!dent)
                        break;
		r = parse_crontab(strappend(CRONTAB_DIR_PATH, dent->d_name));
		if (r < 0) {
                        log_error("%m");
                        goto exit;
                }
	}

exit:
        return (r < 0) ? EXIT_FAILURE : EXIT_SUCCESS;
}


remove_stale_stamps.c
#define _GNU_SOURCE
#include <dirent.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdio.h>

#define TIMERS_DIR "/var/lib/systemd/timers"

int main(int argc, char *argv[]) {
        DIR *dirp;
        struct dirent *dent;
        struct stat sb;
        // dent->d_name[256]
        char basename[170];
        char unit[200];

        if(chdir(TIMERS_DIR)) {
                return 0;
        }

        dirp = opendir(TIMERS_DIR);
        if (dirp == NULL) {
                return 0;
        }

        while ((dent = readdir(dirp))) {
                if (strncmp(dent->d_name, "stamp-cron-", strlen("stamp-cron-")))
                        continue;
                if (strcmp(dent->d_name + strlen(dent->d_name) - strlen(".timer"),".timer"))
                        continue;
		strncpy(basename, &dent->d_name[strlen("stamp-")], sizeof(basename)-1);

                snprintf(unit, sizeof(unit), "/usr/lib/systemd/system/%s", basename);
                if (stat(unit, &sb) != -1)
                        continue;

                snprintf(unit, sizeof(unit), "/run/systemd/generator/%s", basename);
                if (stat(unit, &sb) != -1)
                        continue;

                printf("Removing stale stamp /var/lib/systemd/timers/%s\n", dent->d_name);
                if(unlink(dent->d_name)) {
                     perror("failed");
                };
	}
        closedir(dirp);
}


systemd-cron-generator.c
/*-*- Mode: C; c-basic-offset: 8; indent-tabs-mode: nil -*-*/

/***
  This file is part of systemd.

  Copyright 2013 Shawn Landden

  systemd is free software; you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation; either version 2.1 of the License, or
  (at your option) any later version.

  systemd is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with systemd; If not, see <http://www.gnu.org/licenses/>.
***/

#define _GNU_SOURCE
#include <errno.h>
#include <stdio.h>
#include <unistd.h>
#include <dirent.h>
#include <stdbool.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include <bsd/md5.h>

#ifndef USER_CRONTABS
#define USER_CRONTABS "/var/spool/cron/crontabs"
#endif

static const char *arg_dest = "/tmp";

const char *daysofweek[] = {"Sun","Mon","Tue","Wed","Thu","Fri","Sat"};
const char *isdow = "0123456";

static int parse_dow(char *dow, char *dows){
        char c[2] = {"\0\0"};

        dows[0] = '\0';

        for(unsigned s = 0;s < strlen(dow);s++) {
                c[0] = dow[s];
                if (strchr(isdow, c[0]))
                        strncpy(dows + strlen(dows), daysofweek[atoi(&c[0])], 128 - strlen(dows));
                else if (c[0] == '*')
                        continue;
                else {
                        int l = strlen(dows);
                        if (l < 128) {
                                dows[l] = c[0];
                                dows[l+1] = '\0';
                        }
                }
        }

        int l = strlen(dows);
        if (l) {
                dows[l] = ' ';
                dows[l+1] = '\0';
        }

        return 0;
}

struct text_dict
{
    char *key;
    char *val;
    struct text_dict *next;
};
typedef struct text_dict env;

struct int_dict
{
    char *key;
    int val;
    struct int_dict *next;
};
typedef struct int_dict sequence;

static int parse_crontab(const char *dirname, const char *filename, char *usertab) {
        char *fullname;
        asprintf(&fullname, "%s/%s", dirname, filename);

        FILE *fp = NULL;
        char line[1024];
        char *p;

        char frequency[11], m[25], h[25], dom[25], mon[25], dow[25], user[65];
        char dows[128];
        char *schedule;
        bool persistent = false;

        int remainder = 0;
        int remainder2 = 0;
        char *command;

        /* fake regexp */
        char *pos_equal;
        char *pos_blank;

        env *head = NULL;
        env *curr = NULL;

        /* out */
        sequence *seq_head = NULL;
        sequence *seq_curr = NULL;
        FILE *outp = NULL;
        char *unit = NULL;
        char *outf = NULL;

        fp = fopen(fullname, "r");
        if (!fp) {
            free(fullname);
            return -errno;
        }

        while (fgets(line, sizeof(line), fp)) {
                p = strchr(line, '\n');
                p[0] = '\0';
                switch(line[0]) {
                    case '\0':
                      continue;
                    case '#':
                      continue;
                    case '@':
                      sscanf(line, "%10s %n", frequency, &remainder);
                      strcpy(m, "*");
                      strcpy(h, "*");
                      strcpy(dom, "*");
                      strcpy(mon, "*");
                      strcpy(dow, "*");
                      switch(frequency[1]) {
                          case 'y': // yearly
                          case 'a': // annually
                            mon[0] = '1';
                          case 'm': // monthly
                            dom[0] = '1';
                          case 'w': // weekly
                            if (frequency[1] == 'w')
                                dow[0] = '1';
                          case 'd': // daily
                            h[0] = '0';
                          case 'h': // hourly
                            m[0] = '0';
                      }
                      break;
                    default:
                      pos_equal=strchr(line, '=');
                      pos_blank=strchr(line, ' ');
                      if ((pos_equal != NULL) &&
                          (pos_blank == NULL || pos_equal < pos_blank)) {
                          pos_equal[0]='\0';
                          char *value=pos_equal+1;

                          // lstrip
                          while (value[0] == '"' || value[0] == '\''){
                              value++;
                          }

                          // rstrip
                          for(int i=strlen(value)-1; i>0; i--) {
                              if (value[i] == '"' || value[i] == '\'')
                                  value[i] = '\0';
                              else
                                  break;
                          }

                          if(strcmp("PERSISTENT", line) == 0) {
                              for (int i=0; value[i]; i++)
                                   value[i] = tolower((unsigned char)value[i]);
                              persistent = (!strcmp(value,"true") ||
                                            !strcmp(value,"yes") ||
                                            !strcmp(value,"1"));
                              continue;
                          }

                          curr = head;
                          bool found = false;
                          while(curr) {
                              if (strcmp(curr->key, line) == 0) {
                                  curr->val = (char *)realloc(curr->val, strlen(value)+1);
                                  strcpy(curr->val, value);
                                  found = true;
                                  break;
                              }
                              curr = curr->next;
                          }
                          if (!found) {
                              curr = (env *)malloc(sizeof(env));
                              curr->key = (char *)malloc(strlen(line)+1);
                              strcpy(curr->key, line);
                              curr->val = (char *)malloc(strlen(value)+1);
                              strcpy(curr->val, value);
                              curr->next = head;
                              head = curr;
                          }
                          continue;
                      }
                      sscanf(line, "%24s %24s %24s %24s %24s %n", m, h, dom, mon, dow, &remainder);
                }
                if (usertab == NULL)
                    sscanf(&line[remainder], "%64s %n", user, &remainder2);
		else
                    strcpy(user, usertab);
                command = &line[remainder+remainder2];

                parse_dow(dow, &dows[0]);
                asprintf(&schedule, "%s*-%s-%s %s:%s:00", dows, mon, dom, h, m);
                if (persistent) {
                    unsigned char digest[16];
                    MD5_CTX context;
                    MD5Init(&context);
                    MD5Update(&context, (unsigned char *)schedule, strlen(schedule)+1);
                    MD5Update(&context, (unsigned char *)command, strlen(command));
                    MD5Final(digest, &context);
                    char md5[33];
                    for(int i = 0; i < 16; ++i)
                        sprintf(&md5[i*2], "%02x", (unsigned int)digest[i]);
                    asprintf(&unit, "cron-%s-%s-%s", filename, user, md5);
                } else {
                    seq_curr = seq_head;
                    bool found = false;
                    while(seq_curr) {
                        if (strcmp(seq_curr->key, user) == 0) {
                            seq_curr->val++;
                            found = true;
                            break;
                        }
                        seq_curr = seq_curr->next;
                    }
                    if (!found) {
                        seq_curr = (sequence *)malloc(sizeof(sequence));
                        seq_curr->key = (char *)malloc(strlen(user)+1);
                        strcpy(seq_curr->key, user);
                        seq_curr->val = 0;
                        seq_curr->next = seq_head;
                        seq_head = seq_curr;
                    }
                    asprintf(&unit, "cron-%s-%s-%d", filename, user, seq_curr->val);
                }

                asprintf(&outf, "%s/%s.timer", arg_dest, unit);
                outp = fopen(outf, "w");
                fputs("[Unit]\n", outp);
                fprintf(outp, "Description=[Timer] \"%s\"\n", line);
                fputs("Documentation=man:systemd-crontab-generator(8)\n", outp);
                fputs("PartOf=cron.target\n", outp);
                fputs("RefuseManualStart=true\n", outp);
                fputs("RefuseManualStop=true\n", outp);
                fprintf(outp, "SourcePath=%s\n\n", fullname);

                fputs("[Timer]\n", outp);
                fprintf(outp, "Unit=%s.service\n", unit);
                fprintf(outp, "OnCalendar=%s\n", schedule);
                if (persistent)
                    fputs("Persistent=true\n", outp);
                fclose(outp);

                free(outf);
                asprintf(&outf, "%s/%s.service", arg_dest, unit);
                outp = fopen(outf, "w");
                fputs("[Unit]\n", outp);
                fprintf(outp, "Description=[Cron] \"%s\"\n", line);
                fputs("Documentation=man:systemd-crontab-generator(8)\n", outp);
                fputs("RefuseManualStart=true\n", outp);
                fputs("RefuseManualStop=true\n", outp);
                fprintf(outp, "SourcePath=%s\n\n", fullname);

                fputs("[Service]\n", outp);
                fputs("Type=oneshot\n", outp);
                fputs("IgnoreSIGPIPE=false\n", outp);
                if (strcmp(user, "root")) {
                    fprintf(outp, "User=%s\n", user);
                }
                fprintf(outp, "ExecStart=/bin/sh -c \"%s\"\n", command);

                if (head) {
                    fputs("Environment=", outp);
                    curr = head;
                    while(curr) {
                        if (strlen(curr->val) == 0)
                            {}
                        else if (strchr(curr->val, ' '))
                            fprintf(outp, "%s=\"%s\" ", curr->key, curr->val);
                        else
                            fprintf(outp, "%s=%s ", curr->key, curr->val);
                        curr = curr->next;
                    }
                    fputs("\n", outp);
                }

                fclose(outp);

                free(schedule);
                free(unit);
                free(outf);
        }
        free(fullname);
        fclose(fp);

        curr = head;
        while(curr) {
                free(curr->key);
                free(curr->val);
                head = curr->next;
                free(curr);
                curr = head;
        }

        seq_curr = seq_head;
        while(seq_curr) {
                free(seq_curr->key);
                seq_head = seq_curr->next;
                free(seq_curr);
                seq_curr = seq_head;
        }

        return 0;
}

int parse_dir(bool system, const char *dirname) {
	DIR *dirp;
	struct dirent *dent;

	dirp = opendir(dirname);
	if (dirp == NULL) {
		return 0;
	}

	while ((dent = readdir(dirp))) {
                if (dent->d_name[0] == '.') // '.', '..', '.placeholder'
                    continue;
                if (system)
                    parse_crontab(dirname, dent->d_name, NULL);
                else
                    parse_crontab(dirname, dent->d_name, dent->d_name);
	}
        closedir(dirp);
        return 0;
}

int main(int argc, char *argv[]) {
        if (argc > 1)
                arg_dest = argv[1];

        umask(0022);
        parse_crontab("/etc", "crontab", NULL);
        parse_dir(true, "/etc/cron.d");
        parse_dir(false, USER_CRONTABS);

        return 0;
}
